# Practical Application III: Comparing Classifiers

**Overview**: In this practical application, your goal is to compare the performance of the classifiers we encountered in this section, namely K Nearest Neighbor, Logistic Regression, Decision Trees, and Support Vector Machines.  We will utilize a dataset related to marketing bank products over the telephone.  



### Getting Started

Our dataset comes from the UCI Machine Learning repository [link](https://archive.ics.uci.edu/ml/datasets/bank+marketing).  The data is from a Portugese banking institution and is a collection of the results of multiple marketing campaigns.  We will make use of the article accompanying the dataset [here](CRISP-DM-BANK.pdf) for more information on the data and features.



### Problem 1: Understanding the Data

To gain a better understanding of the data, please read the information provided in the UCI link above, and examine the **Materials and Methods** section of the paper.  How many marketing campaigns does this data represent?

---
#### Answer: The data collected comes from 17 marketing campaigns!
---

### Problem 2: Read in the Data

Use pandas to read in the dataset `bank-additional-full.csv` and assign to a meaningful variable name.

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('data/bank-additional-full.csv', sep = ';')

In [4]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

### Problem 3: Understanding the Features


Examine the data description below, and determine if any of the features are missing values or need to be coerced to a different data type.


```
Input variables:
# bank client data:
1 - age (numeric)
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5 - default: has credit in default? (categorical: 'no','yes','unknown')
6 - housing: has housing loan? (categorical: 'no','yes','unknown')
7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:
8 - contact: contact communication type (categorical: 'cellular','telephone')
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
# other attributes:
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14 - previous: number of contacts performed before this campaign and for this client (numeric)
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
# social and economic context attributes
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
17 - cons.price.idx: consumer price index - monthly indicator (numeric)
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')
```



In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

### Problem 4: Understanding the Task

After examining the description and data, your goal now is to clearly state the *Business Objective* of the task.  State the objective below.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

#### Answer: The business objective is to improve the bank's marketing strategy by promoting term deposits. By leveraging the existing data, we can build a predictive model to determine which customers are more likley to subscribe to a term deposit.

### Problem 5: Engineering Features

Now that you understand your business objective, we will build a basic model to get started.  Before we can do this, we must work to encode the data.  Using just the bank information features, prepare the features and target column for modeling with appropriate encoding and transformations.

In [7]:
# drop the duration column since we wouldn't know this info until after the calls were made
df = df.drop(columns=['duration'])

In [8]:
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
from sklearn.pipeline import Pipeline

categorical_cols = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']
numerical_cols = ['age', 'campaign', 'pdays', 'previous', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed']

onehot_encoder = OneHotEncoder(drop='first', sparse_output=False)
scaler = StandardScaler()

column_transformer = ColumnTransformer(transformers=[
    ('onehot', onehot_encoder, categorical_cols),
    ('scaler', scaler, numerical_cols)
])

label_encoder = LabelEncoder()
df['y'] = label_encoder.fit_transform(df['y'])

pipeline = Pipeline(steps=[
    ('preprocessor', column_transformer)
])

X = pipeline.fit_transform(df)
X = pd.DataFrame(X, columns=pipeline.named_steps['preprocessor'].get_feature_names_out())

In [9]:
y = df['y']

### Problem 6: Train/Test Split

With your data prepared, split it into a train and test set.

In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print(f"Training set size: {X_train.shape}")
print(f"Test set size: {X_test.shape}")


Training set size: (28831, 52)
Test set size: (12357, 52)


### Problem 7: A Baseline Model

Before we build our first model, we want to establish a baseline.  What is the baseline performance that our classifier should aim to beat?

In [11]:
majority_class_proportion = y.value_counts().max() / len(y)
majority_class_proportion

0.8873458288821987

Our baseline accuracy is 88.7%, which is classifying all datapoints as the most commonly occuring class ('no'/0).
Any model we train should at least beat this baseline on the accuracy score or it would not be very good at its job!

### Problem 8: A Simple Model

Use Logistic Regression to build a basic model on your data.  

In [12]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import time

log_reg = LogisticRegression(max_iter=10000, random_state=42)

start_time = time.time()
log_reg.fit(X_train, y_train)
end_time = time.time()

y_train_pred = log_reg.predict(X_train)
y_test_pred = log_reg.predict(X_test)


### Problem 9: Score the Model

What is the accuracy of your model?

In [13]:
log_reg_training_time = end_time - start_time
log_reg_train_accuracy = accuracy_score(y_train, y_train_pred)
log_reg_test_accuracy = accuracy_score(y_test, y_test_pred)

print(f"Training Time: {log_reg_training_time:.4f} seconds")
print(f"Training Accuracy: {log_reg_train_accuracy:.4f}")
print(f"Testing Accuracy: {log_reg_test_accuracy:.4f}")

results_df = pd.DataFrame(columns=['Model', 'Training Time', 'Training Accuracy', 'Testing Accuracy'])
results_df.loc[len(df)] = ['LogisticRegression', log_reg_training_time, log_reg_train_accuracy, log_reg_test_accuracy]

Training Time: 0.3620 seconds
Training Accuracy: 0.9005
Testing Accuracy: 0.9009


### Problem 10: Model Comparisons

Now, we aim to compare the performance of the Logistic Regression model to our KNN algorithm, Decision Tree, and SVM models.  Using the default settings for each of the models, fit and score each.  Also, be sure to compare the fit time of each of the models.  Present your findings in a `DataFrame` similar to that below:

| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|     |    |.     |.     |

In [14]:
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
import time

models = {
    'KNN': KNeighborsClassifier(),
    'Decision Tree': DecisionTreeClassifier(),
    'SVM': SVC()
}

for name, model in models.items():
    X_train = np.array(X_train)
    X_test = np.array(X_test)

    start_time = time.time()
    model.fit(X_train, y_train)
    end_time = time.time()

    train_accuracy = model.score(X_train, y_train)
    test_accuracy = model.score(X_test, y_test)
    
    results_df.loc[len(results_df)] = [name, end_time - start_time, train_accuracy, test_accuracy]

results_df.set_index('Model',inplace=True)

In [15]:
results_df

Training Time  Training Accuracy  Testing Accuracy
Model                                                                 
LogisticRegression       0.361999           0.900489          0.900947
KNN                      0.003999           0.913218          0.892288
Decision Tree            0.216999           0.995595          0.834426
SVM                     47.797999           0.905345          0.900947

### Problem 11: Improving the Model

Now that we have some basic models on the board, we want to try to improve these.  Below, we list a few things to explore in this pursuit.

- More feature engineering and exploration.  For example, should we keep the gender feature?  Why or why not?
- Hyperparameter tuning and grid search.  All of our models have additional hyperparameters to tune and explore.  For example the number of neighbors in KNN or the maximum depth of a Decision Tree.  
- Adjust your performance metric

In [18]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import roc_auc_score, f1_score
import numpy as np
import time

param_grids = {
    'KNN': {'n_neighbors': [3, 5, 7, 9]},
    'Decision Tree': {'max_depth': [None, 10, 20, 30]},
    'SVM': {'kernel': ['linear', 'poly', 'rbf', 'sigmoid']}
}

improved_results_df = pd.DataFrame(columns=['Model', 'Time', 'Train Accuracy', 'Test Accuracy', 'AUC', 'F1'])

# Hyperparameter tuning and model evaluation
for name, model in models.items():
    grid = GridSearchCV(model, param_grids[name], cv=3, n_jobs=-1)

    start_time = time.time()
    grid.fit(X_train, y_train)
    end_time = time.time()

    best_model = grid.best_estimator_
    
    train_accuracy = best_model.score(X_train, y_train)
    test_accuracy = best_model.score(X_test, y_test)
    
    y_pred = best_model.predict(X_test)
    auc_score = roc_auc_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    
    improved_results_df.loc[len(improved_results_df)] = [name, end_time - start_time, train_accuracy, test_accuracy, auc_score, f1]

improved_results_df.set_index('Model', inplace=True)

In [19]:
improved_results_df

Time  Train Accuracy  Test Accuracy       AUC        F1
Model                                                                       
KNN              4.537998        0.907738       0.894716  0.621277  0.877614
Decision Tree    1.611648        0.918733       0.895929  0.627305  0.879633
SVM            167.833879        0.905345       0.900947  0.620700  0.881513

##### Questions